### PR TITLE
Fixes price estimation.

### LINKED ISSUE #950: External prices used in the driver are incorrect.
In this [instance](http://gnosis-europe-dev-gpv2-solver.s3-eu-central-1.amazonaws.com/data/dev/2021/08/06/13%3A23%3A10.517679630_UTC_Ethereum_1/00_instance_raw.json) collected today, the external price of SNX was given as 0.37783117020632184 WETH. The price of SNX today according to trading view is 0.0037 WETH. 

This is more serious than it looks. It is not just the value passed to the MIP solver that is wrong (we're even currently discarding it), but more importantly, this external price is used in the driver to rank solutions. For this particular instance, the winning solution was wrongly the NaiveSolver one, settling 2 orders instead of potentially 12 by quasimodo. 

The surplus [reported](https://logs.gnosisdev.com/app/kibana#/discover/doc/de1f5360-a9ae-11ea-acbb-9d2e3cb124ca/logstash-2021.08.06?id=HJSiG3sB4_N2vXHsD_9I) by the NaiveSolver does not match the surplus I get using the MIP solver to solve the same 2 order instance (it differs by an order of magnitude).

### PR DESCRIPTION
**Pull request by marcovc**
This PR:

* Adds support for estimating prices with or without taking gas costs into account.

* Removes support for computing spot prices. 

Because they are evil, and it was only being used from price_estimate.rs to circumvent a chicken-and-egg problem caused by the need to select the best path by taking gas cost into consideration, and also on unit tests. The bullet above addresses the first issue, while the unit tests were also adapted to test what is actually being used.

* Adds an "amount-to-estimate-prices-with" parameter that should be set to a reasonable amount (like 1 ETH on mainnet or 1000 DAI on XDAI).

Now both estimate_price and estimate_prices will use this value, the former to integrate gas costs, and the latter to bypass the small liquidity paths which cause price estimates to be off in the first place.
 
### Test Plan

Added a unit test, and checked manually if prices now make sense by comparing them to uniswap v2.

Closes #950

### CODE REVIEW

File: shared/src/arguments.rs [hunk 1]
```diff
@@ -88,6 +88,15 @@ pub struct Arguments {
         parse(try_from_str = duration_from_seconds),
     )]
     pub block_stream_poll_interval_seconds: Duration,
+
+    /// Default is one atom of native_token (i.e. price estimation uses spot prices).
```
[the remark this answers is not in the document]
Comment (marcovc) [reply shared/src/arguments.rs:92]: Here's the merge request: https://gitlab.gnosisdev.com/devops/aws/aws-infra-staging/k8s-team-namespaces/gnosis-staging-k8s-gp/-/merge_requests/29

**Comment by marcovc** [comment]
I've merged the new env var to staging. Can I please get a review so that this can be merged and closed?

**Review by vkgnosis — APPROVED** [review]
I still find the price estimator quite hairy so I'm not super confident in my review but I do think it is correct.  I have some ideas on how to refactor this file to get rid of async_recursion and boxing and maybe make it easier to read but I don't want to block this PR with that discussion so we can test in staging earlier.

**Review by nlordell — COMMENTED** [review]
I think it looks correct. I'm not a fan of needing a `proc_macro` to allow recursion for a method that isn't actually recursive (it calls itself, yes, but it "recurses" to a deterministic depth/calltree everytime), so I expect it doesn't actually need to be recursive.

>  I have some ideas on how to refactor this file to get rid of async_recursion and boxing and maybe make it easier to read but I don't want to block this PR with that discussion so we can test in staging earlier.

Sounds like a good plan, seeing as this is important to compare solvers correctly. Lets try and tackle this techdebt as soon as possible. Created https://github.com/gnosis/gp-v2-services/issues/1041.

**Review by nlordell — APPROVED** [review]

**Review by josojo — APPROVED** [review]
Yes, agree with previous reviews. 
But logic looks correct for me as well.

**Pull request merged**